app/RunTest.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE OverloadedStrings #-}
+module RunTest
+  ( findAllSourceFiles
+  , compileTarget
+  , runServer
+  , prettyPrintDiags
+  )
+where
+
+import           GhcMonad
+import           GHC
+import           Control.Monad
+import qualified Control.Concurrent.STM        as STM
+import           Data.List                      ( isPrefixOf )
+import qualified Data.Text                     as T
+import qualified Data.Map                      as Map
+import           Data.Default
+import           System.Directory               ( doesDirectoryExist
+                                                , listDirectory
+                                                , canonicalizePath
+                                                , doesFileExist
+                                                )
+import           System.FilePath
+import           Language.Haskell.LSP.Core
+import           Language.Haskell.LSP.Types
+import           Haskell.Ide.Engine.PluginsIdeMonads
+                                         hiding ( withIndefiniteProgress
+                                                , withProgress
+                                                )
+import           Haskell.Ide.Engine.GhcModuleCache
+import qualified Haskell.Ide.Engine.ModuleCache
+                                               as MC
+import qualified Haskell.Ide.Engine.Ghc        as Ghc
+
+findAllSourceFiles :: FilePath -> IO [FilePath]
+findAllSourceFiles fp = do
+  absFp <- canonicalizePath fp
+  isDir <- doesDirectoryExist absFp
+  if isDir
+    then findFilesRecursively
+      isHaskellSource
+      (\path -> any (\p -> p path) [isHidden, isSpecialDir])
+      absFp
+    else filterM doesFileExist [absFp]
```

Comment:
silently discarding unknown filepaths does not seem pretty, but neither does throwing an exception.